### PR TITLE
Visualization 

### PR DESCRIPTION
**Pull request by SergiiLisnyi**

### CODE REVIEW

File: UITable/UITable/ActionTextField.swift [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+//
+
+import Foundation
+import UIKit
+
+class ActionTextField: UITextField, UITextFieldDelegate {
+    
+    var action: ((String)->())?
+    
+    init(title: String, action: @escaping (String)->()) {
```
Comment (oleksii-lubianyi-epam) [inline UITable/UITable/ActionTextField.swift]: Where do you use `title`?

File: UITable/UITable/ActionTextField.swift [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    //MARK: TextFieldDelegate
+    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
+        if let text = textField.text as NSString? {
+            let txtAfterUpdate = text.replacingCharacters(in: range, with: string)
+
+            //limit count
+            if txtAfterUpdate.count > 3 {
+                return false
+            }
+            action!(txtAfterUpdate)
```
Comment (oleksii-lubianyi-epam) [inline UITable/UITable/ActionTextField.swift]: `action` is optional. Force unwrap is not safe.
Either use `action?.(txtAfterUpdate)`, or declare it non-optional or force unwrapped

File: UITable/UITable/ActionTextField.swift [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    //MARK: TextFieldDelegate
+    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
+        if let text = textField.text as NSString? {
+            let txtAfterUpdate = text.replacingCharacters(in: range, with: string)
+
+            //limit count
+            if txtAfterUpdate.count > 3 {
```
Comment (oleksii-lubianyi-epam) [inline UITable/UITable/ActionTextField.swift]: Don't use "magic numbers". 
Store them in `let` constants with descriptive names.

File: UITable/UITable/Adapter.swift [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    
+    func place(control: ATDControlProtocol, view: UIStackView){
+        let arr: [TypeDate] = control.createMenu()
+       
+        for typeDate in arr {
+            switch typeDate {
+            case .button(let title, let action):
+                view.addArrangedSubview(ActionButton(title: title, action: action))
+        
+            case .textField(let title, let action):
```
Comment (oleksii-lubianyi-epam) [inline UITable/UITable/Adapter.swift]: UITextfiled has either `text` or `placeholder` properties, not `title`

File: UITable/UITable/Adapter.swift [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+//  Created by Sergii Lisnyi on 4/27/18.
+//  Copyright © 2018 Sergii Lisnyi. All rights reserved.
+//
+
+import Foundation
+import  UIKit
+
+class Adapter {
+    
+    func place(control: ATDControlProtocol, view: UIStackView){
```
Comment (oleksii-lubianyi-epam) [inline UITable/UITable/Adapter.swift]: You are not placing `control` anywhere. Think about better naming

File: UITable/UITable/ControlManagerFactory.swift [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+//  Created by Sergii Lisnyi on 4/26/18.
+//  Copyright © 2018 Sergii Lisnyi. All rights reserved.
+//
+
+import Foundation
+
+class ControlManagerFactory  {
+    
+    static func getManager(model: EntityProtocol) -> ATDControlProtocol {
+        let title = model.name
```
Comment (oleksii-lubianyi-epam) [inline UITable/UITable/ControlManagerFactory.swift]: Replace `model.name` with `model.type`, where `type` is enumeration

File: UITable/UITable/Controller/DetailController.swift [hunk 7]
```diff
@@ -17,24 +17,17 @@ class DetailController: UIViewController {
...
-        viewForOpacity.opacityGradient()
-    }
-
-    override func viewDidAppear(_ animated: Bool) {
-        super.viewDidAppear(true)
-    }
-    
-    override func didReceiveMemoryWarning() {
-        super.didReceiveMemoryWarning()
+        textLabel.text = model.descr
```
Comment (oleksii-lubianyi-epam) [inline UITable/UITable/Controller/DetailController.swift]: Don't shorten words in properties names (until it's really required).
`description` is` NSObject` property, so think about better name.

File: UITable/UITable/Controller/DetailController.swift [hunk 8]
```diff
@@ -66,37 +59,70 @@ class DetailController: UIViewController {
     
     @IBAction func openWikiButtonTapped(_ sender: UIButton) {
         let alert = UIAlertController(title: "Choose browser", message: nil, preferredStyle: .actionSheet)
-        alert.addAction(getAction(title: "UIWebView", idController: "uiWebID"))
-        alert.addAction(getAction(title: "WKWebView", idController: "wkWebID"))
-        alert.addAction(getAction(title: "SFSafariView", idController: "sfWebID"))
+        alert.addAction(getAction(title: "UIWebView", controllerID: "uiWebID"))
```
Comment (oleksii-lubianyi-epam) [inline UITable/UITable/Controller/DetailController.swift]: You already have `ActionType. titleAndControllerID` - why not use them?
Also, I am not sure if you really need that `And`. I would rather suggest to split `title` an `controllerId` into two different properties.

File: UITable/UITable/Controller/DetailController.swift [hunk 9]
```diff
@@ -66,37 +59,70 @@ class DetailController: UIViewController {
...
         let gradient = CAGradientLayer()
         gradient.frame = self.bounds
         gradient.colors = [UIColor.clear.cgColor, UIColor.black.cgColor, UIColor.black.cgColor, UIColor.black.cgColor]
         gradient.locations = [0 , 1]
         self.layer.mask = gradient
     }
 }
+
+extension UIButton {
+    func setTitle(_ title: String) {
```
Comment (oleksii-lubianyi-epam) [inline UITable/UITable/Controller/DetailController.swift]: Is this much better than `setTitle(title, for: .normal)`? I don't think so

File: UITable/UITable/Controller/FakeDataController.swift [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+//  Copyright © 2018 Sergii Lisnyi. All rights reserved.
+//
+
+import UIKit
+
+class FakeDataController: UIViewController, FakeDataProtocol {
+
+    @IBOutlet weak var dataTable: UITableView!
+    
+    var modelData = ModelFakeData()
```
Comment (oleksii-lubianyi-epam) [inline UITable/UITable/Controller/FakeDataController.swift]: If you don't change `modelData` - make it `let`

File: UITable/UITable/Controller/FakeDataController.swift [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        dataTable.separatorColor = UIColor.black
```
Comment (oleksii-lubianyi-epam) [inline UITable/UITable/Controller/FakeDataController.swift]: Move into storyboard

File: UITable/UITable/Controller/FakeDataController.swift [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    }
+    
+    func deleteToIndex(index: Int) {
+        guard index < modelData.count  else { print ("Index out of range"); return }
+        modelData.remove(at: index)
+        dataTable.deleteRows(at: [IndexPath(row: index, section: 0)], with: .automatic)
+    }
+    
+    func commit() {
+        for i in 0..<modelData.count{
```
Comment (oleksii-lubianyi-epam) [inline UITable/UITable/Controller/FakeDataController.swift]: Think about using `foreach`

File: UITable/UITable/Controller/FakeDataController.swift [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+extension FakeDataController: UITableViewDelegate, UITableViewDataSource {
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        
+        guard let cell = tableView.dequeueReusableCell(withIdentifier: DataTableCell.identifier, for: indexPath) as? DataTableCell else  {
+            return UITableViewCell()
+        }
+        cell.label?.text = modelData.getValue(at: indexPath.row) + modelData.getHelperValue(at: indexPath.row)
```
Comment (oleksii-lubianyi-epam) [inline UITable/UITable/Controller/FakeDataController.swift]: `DataTableCell.label` declared as force unwrapped. Why `?`

File: UITable/UITable/Controller/FakeDataController.swift [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+        modelData.getState(at: indexPath.row) ? (cell.label.backgroundColor = UIColor.clear) : (cell.label.backgroundColor = UIColor.red)
```
Comment (oleksii-lubianyi-epam) [inline UITable/UITable/Controller/FakeDataController.swift]: Add `color` property into model and calculate it there, not in controller

File: UITable/UITable/Controller/SFSafariVC.swift [hunk 15]
```diff
@@ -13,17 +13,22 @@ class SFSafariVC: UIViewController, SFSafariViewControllerDelegate, WebProtocol
     
     var getLink: String?
     var vc: SFSafariViewController!
+    var isDone = false
```
Comment (oleksii-lubianyi-epam) [inline UITable/UITable/Controller/SFSafariVC.swift]: Should `getLink` (above) be optional?

File: UITable/UITable/Controller/VisualizationController.swift [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+//  Copyright © 2018 Sergii Lisnyi. All rights reserved.
+//
+
+import UIKit
+
+class VisualizationController: UIViewController {
+
+    @IBOutlet weak var stackForButtons: UIStackView!
+    weak var tableController: FakeDataController?
+    var adapter = Adapter()
```
Comment (oleksii-lubianyi-epam) [inline UITable/UITable/Controller/VisualizationController.swift]: `let`?

File: UITable/UITable/DataStructureViewCell.swift [hunk 17]
```diff
@@ -25,11 +21,9 @@ class DataStructureViewCell: UITableViewCell {
     }
 }
 
-
 extension DataStructureViewCell: ConfigureCellProtocol {
     
     func configure(model: EntityProtocol) {
-        //label.text = model.currentDataStruct?.name
-        label.text = model.getName()
+        label.text = model.name
```
Comment (oleksii-lubianyi-epam) [inline UITable/UITable/DataStructureViewCell.swift]: `configureWith` (above)

File: UITable/UITable/DataTableCell.swift [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+    
+    static let identifier = "customCell"
+    
+    @IBOutlet weak var label: UILabel!
+    override func awakeFromNib() {
+        super.awakeFromNib()
+        // Initialization code
+    }
+
+    func setName(name: Int) {
```
Comment (oleksii-lubianyi-epam) [inline UITable/UITable/DataTableCell.swift]: Controller (or, better, view model) should prepare presentation data.
Cell should not perform transformation

File: UITable/UITable/Protocol/ATDControlProtocol.swift [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+
+import Foundation
+
+protocol ATDControlProtocol {
+    
+    weak var delegate: FakeDataProtocol? {get set}
+    
+    func createMenu() -> [TypeDate]
+    
+    func setDelegate(delegate: FakeDataProtocol)
```
Comment (oleksii-lubianyi-epam) [inline UITable/UITable/Protocol/ATDControlProtocol.swift]: Remove `func setDelegate(delegate: FakeDataProtocol)` and it's usage - `weak var delegate: FakeDataProtocol? {get set}` is enough

File: UITable/UITable/Manager/MultiSetManager.swift [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+class MultiSetManager: ATDControlProtocol {
+    
+    var delegate: FakeDataProtocol?
+    var value = ""
+    
+    func setDelegate(delegate: FakeDataProtocol) {
+        self.delegate = delegate
+    }
+    
+    func change(_ value: String) {
```
Comment (oleksii-lubianyi-epam) [inline UITable/UITable/Manager/MultiSetManager.swift]: Why not call `self.value = <smth>` directly?

File: UITable/UITable/Manager/MultiSetManager.swift [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+        self.value = value
+    }
+    
+    func plusNumberOfRepetitions(_ value: String) -> String {
+        guard value.count != 0  else { return " - 2 times"}
+            let repeatition = Int(value.components(separatedBy: CharacterSet.decimalDigits.inverted).joined())
```
Comment (oleksii-lubianyi-epam) [inline UITable/UITable/Manager/MultiSetManager.swift]: `if let` to avoid force unwrap

File: UITable/UITable/Manager/MultiSetManager.swift [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+            return " - \(repeatition! + 1) times"
+    }
+    
+    func minusNumberOfRepetitions(_ value: String) -> String {
+        guard value != " - 2 times" else { return ""}
+        let repeatition = Int(value.components(separatedBy: CharacterSet.decimalDigits.inverted).joined())
```
Comment (oleksii-lubianyi-epam) [inline UITable/UITable/Manager/MultiSetManager.swift]: `if let` to avoid force unwrap

**Review by ghost — APPROVED** [review]

**Pull request merged**